### PR TITLE
[CRI] Add TTY flag to AttachRequest

### PR DESCRIPTION
**Pull request by timstclair**
Follow up from https://github.com/kubernetes/kubernetes/pull/35661
For https://github.com/kubernetes/kubernetes/issues/29579

- Add TTY to the CRI AttachRequest
- Moves responsibility from the runtime shim to the Kubelet for populating the TTY bool in the request based on the container spec

/cc @euank @feiskyer @kubernetes/sig-node

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36615)

### CODE REVIEW

**Comment by timstclair** [comment]
Rebased.

File: pkg/kubelet/api/v1alpha1/runtime/api.proto [hunk 1]
```diff
@@ -722,6 +722,8 @@ message AttachRequest {
     optional string container_id = 1;
     // Whether to stream stdin.
     optional bool stdin = 2;
+    // Whether use tty
```
Comment (yujuhong) [inline pkg/kubelet/api/v1alpha1/runtime/api.proto]: nit (just for consistency): missing a period at the end. 
s/Whether use/Whether to

File: pkg/kubelet/kubelet_pods.go [hunk 2]
```diff
@@ -1331,7 +1331,18 @@ func (kl *Kubelet) GetAttach(podFullName string, podUID types.UID, containerName
 			return nil, fmt.Errorf("container not found (%q)", containerName)
 		}
 
-		return streamingRuntime.GetAttach(container.ID, streamOpts.Stdin, streamOpts.Stdout, streamOpts.Stderr)
+		// We need the api.Pod to get the TTY status.
```
Comment (yujuhong) [inline pkg/kubelet/kubelet_pods.go]: nit: could you expand the comment to explain why this is necessary?

File: pkg/kubelet/kubelet_pods.go [hunk 3]
```diff
@@ -1331,7 +1331,18 @@ func (kl *Kubelet) GetAttach(podFullName string, podUID types.UID, containerName
...
+		pod, found := kl.GetPodByFullName(podFullName)
+		if !found || pod.UID != podUID {
+			return nil, fmt.Errorf("pod %s not found", podFullName)
+		}
+		containerSpec := kubecontainer.GetContainerSpec(pod, containerName)
+		if containerSpec == nil {
+			return nil, fmt.Errorf("container not found (%q)", containerName)
```
Comment (yujuhong) [inline pkg/kubelet/kubelet_pods.go]: Very minor, optional nit: the error format can be consistent with the error in line 1337

**Review by yujuhong — COMMENTED** [review]
LGTM with nits.

File: pkg/kubelet/kubelet_pods.go (hunk 3 above)
Comment (timstclair) [inline pkg/kubelet/kubelet_pods.go]: Done.

File: pkg/kubelet/kubelet_pods.go (hunk 2 above)
Comment (timstclair) [inline pkg/kubelet/kubelet_pods.go]: Done.

**Comment by feiskyer** [comment]
LGTM

**Comment by euank** [comment]
Thank you! LGTM

**Comment by feiskyer** [comment]
@yujuhong Add 1.5 milestone for this?

**Comment by euank** [comment]
One nit is the commit title. I didn't realize it was the same as the issue title, but that's ambiguous between "manage a tty for attach" and "set the tty bool".

Clarification so it's obviously the later (suggest: "cri: add tty flag to attach request") would be appreciated!

**Comment by timstclair** [comment]
Done. Squashed & rebased.

**Comment by timstclair** [comment]
Fix gofmt

**Comment by euank** [comment]
I think the commit title got lost in rebase, `Merge pull request #36899 from gmarek/constraints` doesn't look quite right :)

**Comment by timstclair** [comment]
Thanks for catching that! I accidentally squashed it with the commit above :( will fix...

**Comment by feiskyer** [comment]
@yujuhong Add 1.5 milestone for this PR?

**Comment by timstclair** [comment]
@feiskyer - I'd love to get this in, but can't justify it post-freeze. Do you have an argument for why it should be in 1.5?

**Comment by euank** [comment]
I've had this discussion in a different issue as well, this seems like something for post-freeze to me too.

IIRC the freeze on master lifts very soon anyways..

**Comment by feiskyer** [comment]
>  I'd love to get this in, but can't justify it post-freeze. Do you have an argument for why it should be in 1.5?

@timstclair  Just thinking it is bug fix for AttachRequest. But I'm ok moving to next release if you d on't think so.

**Comment by Crazykev** [comment]
@timstclair Could you tell me why missing a tty flag in [Runtime Attach interface](https://github.com/kubernetes/kubernetes/blob/b5b9397291ef46d0b49fc5f7a4b0dfe70a0c8d15/pkg/kubelet/server/streaming/server.go#L61)? I've seen that in [design doc](https://docs.google.com/document/d/1OE_QoInPlVCK9rMAx9aybRmgFiVjHpJCHI9LrfdNM_s/edit#) and didn't see any discussion about it. 

PS: This question may be little distractive, since there is follow up of #35661, I just ask here.

**Comment by timstclair** [comment]
Squashed & rebased.

**Comment by timstclair** [comment]
@Crazykev good catch, that should be updated in this PR.

**Comment by timstclair** [comment]
@Crazykev @yujuhong PTAL

**Comment by yujuhong** [comment]
LGTM

**Comment by Crazykev** [comment]
yeah, that fixed. LGTM

**Comment by timstclair** [comment]
Squashed. Reapplying LGTM per comments.

**Comment by timstclair** [comment]
Trivial rebase, reapplying LGTM.

**Pull request merged**